clash-lib/src/Clash/Normalize/Transformations/Inline.hs

```diff
@@ -379,6 +382,45 @@ inlineCleanup (TransformContext is0 _) (Letrec binds body) = do
 inlineCleanup _ e = return e
 {-# SCC inlineCleanup #-}
 
+-- | Takes a binding and collapses its term if it is a noop
+inlineIdentity :: HasCallStack => NormRewrite
+inlineIdentity _ (Letrec binds body) = do
+  (untouched,collapsed) <- partitionEithers <$> mapM identityAnalysis binds
+  if null collapsed
+    then return (Letrec binds body)
+    else changed (Letrec (untouched ++ collapsed) body)
+inlineIdentity _ e = return e
+
+identityAnalysis :: HasCallStack => (Id,Term) -> NormalizeSession (Either (Id,Term) (Id,Term))
+identityAnalysis old@(iD,collectArgs -> (term,partitionEithers -> (termArgs,typeArgs))) =
+  runIdentityAnalysis $ do
+    Prim primInfo <- return term
+    identity      <- getIdentity primInfo
+    let resTy = polyResTyView $ primType primInfo
+    return (iD,primUCo `App` identity `mkTyApps` (resTy:typeArgs))
+  where
+    runIdentityAnalysis =
+      fmap (Maybe.maybe (Left old) Right) . runMaybeT
+
+    getTermArg i = do
+      Monad.guard $ i <= length termArgs - 1
+      return $ termArgs List.!! i
```

Comment:
Stylistically I'm not a fan of qualified operators. I think it would be better to

```haskell
import Data.List ((!!))
```

at the top of the file